File: 1.0.0(Rust)/src/lexer.rs
use crate::token::Token;
use crate::ast::*;
pub struct Lexer {
    chars: Vec<char>,
    pos: usize,
}

impl Lexer {

    pub fn new(code: String) -> Self {
        Self {
            chars: code.chars().collect(),
            pos: 0,
        }
    }

    fn skip(&mut self) {
        while self.pos < self.chars.len()
            && self.chars[self.pos].is_whitespace()
        {
            self.pos += 1;
        }
    }

    pub fn next_token(&mut self) -> Token {

        self.skip();

        if self.pos >= self.chars.len() {
            return Token::EOF;
        }

        let ch = self.chars[self.pos];

        match ch {

            '(' => { self.pos += 1; Token::LParen }
            ')' => { self.pos += 1; Token::RParen }
            '['=>{self.pos+=1;Token::LBracket}
            ']'=>{self.pos+=1;Token::RBracket}
            '{' => { self.pos += 1; Token::LBrace }
            '}' => { self.pos += 1; Token::RBrace }
            ',' => { self.pos += 1; Token::Comma }
            ';' => { self.pos += 1; Token::Semi }
            '=' => { self.pos += 1; Token::Equal }
            ':' => { self.pos += 1; Token::Colon }
            '-' => { self.pos += 1; Token::Minus }
            '>' => { self.pos += 1; Token::Gt }
            '<' => { self.pos += 1; Token::Lt }
            '*' => { self.pos += 1; Token::Star }
            '&' => { self.pos += 1; Token::And }

            '0'..='9' => {
                let mut n = String::new();
                let mut is_float = false;

                while self.pos < self.chars.len()
                    && (self.chars[self.pos].is_numeric()
                        || self.chars[self.pos] == '.')
                {
                    if self.chars[self.pos] == '.' {
                        is_float = true;
                    }

                    n.push(self.chars[self.pos]);
                    self.pos += 1;
                }

                if is_float {
                    Token::Float(n.parse().unwrap())
                } else {
                    Token::Number(n.parse().unwrap())
                }
            }

            '"' => {
                self.pos += 1;
                let mut s = String::new();

                while self.pos < self.chars.len()
                    && self.chars[self.pos] != '"'
                {
                    s.push(self.chars[self.pos]);
                    self.pos += 1;
                }

                self.pos += 1;
                Token::String(s)
            }

            'a'..='z' | 'A'..='Z' => {
                let mut w = String::new();

                while self.pos < self.chars.len()
                    && self.chars[self.pos].is_alphanumeric()
                {
                    w.push(self.chars[self.pos]);
                    self.pos += 1;
                }
                if let Some(i) = TypeKind::from_string(w.clone().as_str()){
                    return Token::TyKd(i);
                }
                match w.as_str() {
                    "mehen" => Token::Mehen,"frnct" => Token::Func,
                    "urso" => Token::Urso,"erutnos" => Token::Return,
                    "valt" => Token::Valt,"asken" => Token::As,
                    "trunth" => Token::Bool(true),
                    "frunth" => Token::Bool(false),
                    "noph" => Token::Void,
                    _ => Token::Identifier(w),
                }
            }

            _ => panic!("invalid char"),
        }
    }
    
}

File: Python/1.0.5(Python)/cache/lexer.py
# lexer.py
import re
from dataclasses import dataclass

class Token:
    def __init__(self, type, value, line, col, pos):
        self.type = type
        self.value = value
        self.line = line
        self.col = col
        self.pos = pos

TYPES = {
    "intg","flotg","strg","boletg","listh","dicth","aphe",
    "typct_taphlot","typct_neom","veot",
    "kalf","karchen","peontderen",
}

KEYWORDS = {
"kalfen":"KALFEN","frounct":"FRCT","frutupheFrounct":"PFRCT","frutuphenos":"FUTUREC",
"ifnt":"IF","elsnt":"ELSE","efnt":"ELIF",
"peontDelnos":"PTRDEL","forp":"FOR","perl":"PERL","eower":"WHILE",
"keryenos":"FORPPERL","maphnos":"MAPHPERL","asken":"AS",
"serl":"SERL","inkleodnos":"IMPORT",
"kalfenCernos":"CLASSCRE","oed":"YIELD",
}


unary = {"brof":"AND","neat":"NOT","ophe":"OR","asp":"IS","mot":"MOD","intf":"IN"}
TOKEN_SPEC = [
    ("NUMBER",   r"-?\d+\.\d+|-?\d+"),("STRING",   r'"[^"]*"'),
    ("CHAR",     r"'[^']*'"),("BOOL",     r"(trunth|franth)"),("NULL",     r"(noph)"),
    ("EQEQ",     r"=="),("EQUAL",    r"="),

    ("PLUS",     r"\+"),("MINUS",    r"-"),("STAR",     r"\*"),("SLASH",    r"/"),("UP",    r"\^"),
    ("EQLT",     r"<="),("EQGT",     r">="),("LT",       r"<"),("GT",       r">"),("REF",       r"&"),

    ("LPAREN",   r"\("),("RPAREN",   r"\)"),
    
    ("LBRACE",   r"\{"),("RBRACE",   r"\}"),

    ("LBRACKET", r"\["),("RBRACKET", r"\]"),

    ("COMMA",    r","),("COLON",    r":"),("SEMI",     r";"),("POINT",    r"\."),

    ("IDENTIFEN",r"[a-zA-Z_][a-zA-Z0-9_öüİÖÜ]*"),
    ("LABEL",    r'#[^"]*#'),("SKIP",     r"[ \t\n]+"),
]
trns = {
    "IDENTIFEN":"texth",
    "POINT":"detot","COMMA":"vetot",
    "LPAREN":"lofPartot","RPAREN":"roghPartot",
    "LBRACE":"lofSupartot","RBRACE":"roghSupartot",
    "LBRACKET":"lofEpartot","RBRACKET":"roghEpartot",
    "COLON":"fedetot","SEMI":"vedetot","UP":"ipherLitot",
    "GT":"roghLitot","LT":"lofLitot",
    "KALFEN":"kalfen","FRCT":"frounct",
    "IF":"ifnt","ELSE":"elsnt","ELIF":"efnt",
    "PTRDEL":"peontDelnos",
    "MEHEN":"mehen","FOR":"forp","PERL":"perl","SERL":"serl",
    "FORPPERL":"keryenos","MAPHPERL":"maphnos","AS":"asken",
    "TYPE":"typect","EQUAL":"qenev(=)","EQEQ":"duphAfonQenev",
    "EQLT":"kanevOphQenev","EQGT":"banevOphQenev",
    "SLASH":"slatot","PLUS":"adfoss","MINUS":"edfoss","YIELD":"oed",
    "STRING":"sterg","NULL":"nophnev","NUMBER":"neom","STAR":"keush","CHAR":"karch"
}
def trans(a):
    return trns.get(a,"texth")

TOKEN_REGEX = "|".join(f"(?P<{n}>{r})" for n, r in TOKEN_SPEC)
def lex(code: str):
    tokens = []
    line = 1
    col = 0    
    for m in re.finditer(TOKEN_REGEX, code):
        kind = m.lastgroup;value = m.group()
        start_pos = m.start()
        end_pos = m.end()
        # satır ve sütun takibi
        newlines = value.count("\n")
        if newlines > 0:line += newlines;col = len(value) - value.rfind("\n") - 1
        else:col += len(value)
        if kind in ("SKIP","LABEL"):
            continue
        # TYPE ayrımı
        if kind == "IDENTIFEN" and value in TYPES:kind = "TYPE"
        if kind == "IDENTIFEN" and value in KEYWORDS:kind = KEYWORDS[value]
        if kind == "IDENTIFEN" and value in unary:kind = unary[value]
        tokens.append(Token(kind, value, line, col - len(value), (start_pos, end_pos)))
    return tokens